Automated Market Maker (XLS-30d) reference documentation

mDuo13 opened the pull request:
Add reference documentation for [XLS-30d](https://github.com/XRPLF/XRPL-Standards/discussions/78) Automated Market Makers, including:

- [x] New `AMM` ledger entry
- [x] Changes to `AccountRoot` ledger entry (new lsfAMM flag, special rules for AMM AccountRoot objects)
- [x] 5 new AMM transaction types
    - [x] AMMBid
    - [x] AMMCreate
    - [x] AMMDeposit
    - [x] AMMVote
    - [x] AMMWithdraw
- [x] New AMM-related transaction result codes
- [x] New `amm_info` API method
- [x] Update sample JSON with latest round of changes (AMMID→Asset/Asset2, withdraw/deposit flags, etc.)
- [x] Add new "Issue" (asset definition) field to serialization docs

Edit: now ready for review.

These docs have been updated based on the latest code (commit 2fa59f1c68d37ebea4d6d6f098b7378c02251a27) including several breaking changes that have not yet made their way to AMM-Devnet. Reviewers, please be on the lookout for places where the code/functionality may have changed since the previous drafts of these docs.

mDuo13 committed:
AMM reference docs

AMM object: some clarifications

AMM references: incomplete draft

AMM updates

AMM: add AMMBid, clarify more stuff

Docs for AMMBid transaction

AMM: more work on error codes

AMM: example transactions, related upates

Fix broken links

AMM: reflect naming updates

more AMMBid edits

Add amm_info API method

Add style for AMM label

AMM: more clarifications on tx, accounts

AMM references: Clean up draft

amm_info: reflect code updates

AMM: updates for the latest changes

AMM: update flags for Deposit/Withdraw

AMM: update JSON examples for latest code

Serialization: add STIssue type

STIssue: clarify definition with commas

gregtatcam commented on content/references/http-websocket-apis/public-api-methods/path-and-order-book-methods/amm_info.md:
````diff
@@ -0,0 +1,153 @@
+```json
+{
+    "command": "{{currentpage.name}}",
+    "asset1": {
````
It's `asset` for consistency with other fields, which don't include `1`.

gregtatcam commented on content/references/http-websocket-apis/public-api-methods/path-and-order-book-methods/amm_info.md:
```diff
@@ -0,0 +1,153 @@
+{
+    "command": "{{currentpage.name}}",
+    "asset1": {
+      "currency": "XRP"
```
Current implementation enforces a shortcut `"asset": "XRP"`. Amount type has the same shortcut for "XRP" amounts.

mDuo13 replied:
Per our conversation, this is changed back, right?

gregtatcam replied:
Correct.

gregtatcam commented on content/references/http-websocket-apis/public-api-methods/path-and-order-book-methods/amm_info.md:
```diff
@@ -0,0 +1,153 @@
+  "result": {
+    "AMMAccount": "rUKN5XC4S7SRMB8ezRNADr4WWunJJz8jqK",
+    "AMMID": "4B82E4CB90FBE26EA172A498F7A7B03A2A7C285149A9B18731370F2644B96F7A",
+    "Asset1": "250000000",
```
Should be `Asset`.

gregtatcam commented on content/references/http-websocket-apis/public-api-methods/path-and-order-book-methods/amm_info.md:
````diff
@@ -0,0 +1,153 @@
+
+```json
+{
+  "result": {
````
All fields are placed under top level: `"amm": {....}`

gregtatcam commented on content/references/protocol-reference/ledger-data/ledger-object-types/amm.md:
````diff
@@ -0,0 +1,138 @@
+```json
+{
+  "AMMAccount": "rLcJnZS6M9DyZ23g3GPvtyuYgT5QFwWXaN",
+  "AMMToken": {
````
`Token1/2` are moved up from the top-level `AMMToken`. Also renamed to `Asset/Asset2`. There is a new `STIssue` type (borrowed from side-chains). The fields are `currency` and `issuer`.

gregtatcam commented on content/references/protocol-reference/ledger-data/ledger-object-types/amm.md:
```diff
@@ -0,0 +1,138 @@
+| Field            | JSON Type           | [Internal Type][] | Required? | Description  |
+|:-----------------|:--------------------|:------------------|:-------------|
+| `AMMToken`       | Object              | STObject          | Yes       | Defines the two assets in this AMM's pool, as an [AMMToken Object](#ammtoken-object). |
+| `AccountRootID`  | String              | AccountID         | Yes       | The address of the special account that holds this AMM's assets. |
```
It's `AMMAccount`.

gregtatcam commented on content/references/protocol-reference/ledger-data/ledger-object-types/amm.md:
```diff
@@ -0,0 +1,138 @@
+
+### AMMToken Object
+
+The `AMMToken` field is an object that defines the assets in this AMM's pool. Each asset can be XRP or a [fungible token](tokens.html) but they cannot both be XRP.
```
This is removed and the field name changed: `Asset/Asset2`.

gregtatcam commented on content/references/protocol-reference/ledger-data/ledger-object-types/amm.md:
```diff
@@ -0,0 +1,138 @@
+| `AuthAccounts`  | Array               | STArray           | No        | A list of at most 4 additional accounts that are authorized to trade at the discounted fee for this AMM instance. |
+| `DiscountedFee` | String              | UInt32            | Yes       | The trading fee to be charged to the auction owner, in the same format as `TradingFee`. By default this is 0, meaning that the auction owner can trade at no fee instead of the standard fee for this AMM. |
+| `Price`         | [Currency Amount][] | Amount            | Yes       | The amount the auction owner paid to win this slot, in LP Tokens. |
+| `TimeStamp`     | String              | UInt32            | Yes       | The time when this slot was bought, in [seconds since the Ripple Epoch][]. |
```
Changed to `Expiration`.

gregtatcam commented on content/references/protocol-reference/ledger-data/ledger-object-types/amm.md:
```diff
@@ -0,0 +1,138 @@
+      "issuer": "rLcJnZS6M9DyZ23g3GPvtyuYgT5QFwWXaN",
+      "value": "0.7115124735378855"
+    },
+    "TimeStamp": 719029222
```
Changed to `Expiration`.

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,130 @@
+
+    There are two special cases for the cost to outbid someone. In the **first interval** after someone wins the bid, the price to outbid them is only 5% more than their bid:
+
+        P = B × 1.05
```
It's updated to `P = B*1.05 + M`.

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammcreate.md:
```diff
@@ -0,0 +1,72 @@
+{
+  "TransactionType": "AMMCreate",
+  "Account": "rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm",
+  "Asset1": {
```
Changed to `Amount/Amount2`

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
````diff
@@ -0,0 +1,109 @@
+```json
+{
+  "TransactionType": "AMMDeposit",
+  "AMMID": "4B82E4CB90FBE26EA172A498F7A7B03A2A7C285149A9B18731370F2644B96F7A",
````
`AMMID` is replaced with `Asset/Asset2`.

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,109 @@
+  "TransactionType": "AMMDeposit",
+  "AMMID": "4B82E4CB90FBE26EA172A498F7A7B03A2A7C285149A9B18731370F2644B96F7A",
+  "Account": "rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm",
+  "Asset1In": {
```
Changed to `Amount/Amount2`.

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,109 @@
+| `Asset1In`    | [Currency Amount][] | Amount            | No        | The amount of one asset to deposit to the AMM. If present, this must match the type of one of the assets (tokens or XRP) in the AMM's pool. |
+| `Asset2In`    | [Currency Amount][] | Amount            | No        | The amount of another asset to add to the AMM. If present, this must match the type of the other asset in the AMM's pool and cannot be the same asset as `Asset1In`. |
+| `EPrice`      | [Currency Amount][] | Amount            | No        | The maximum effective price, in the deposit asset, to pay for each LP Token received. |
+| `LPTokensOut` | [Currency Amount][] | Amount            | No        | How many of the AMM's LP Tokens to buy. |
```
Changed to `LPTokenOut`.

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,109 @@
+- **Single-asset deposits**, in which you provide only one of the AMM's two assets. You can envision a single-asset deposit as trading _some_ of the deposited asset for the other asset, then performing a double-asset deposit. The trading fee applies to the amount you would need to trade for, but not to the rest of the deposit; the amount of the fee is debited from the LP Tokens paid out. _For exa ...
+
+The following combinations of fields indicate a **double-asset deposit**:
+
```
Flags are required (added below) to specify a valid set of parameters.

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,109 @@
+
+| Fields Specified          | Meaning |
+|---------------------------|---------|
+| `LPTokensOut` only        | Deposit both of this AMM's assets, in amounts calculated so that you receive the specified amount of LP Tokens in return. The amounts deposited maintain the relative proportions of the two assets the AMM already holds. |
```
Flag `tfLPToken`: 0x01

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,109 @@
+| Fields Specified          | Meaning |
+|---------------------------|---------|
+| `LPTokensOut` only        | Deposit both of this AMM's assets, in amounts calculated so that you receive the specified amount of LP Tokens in return. The amounts deposited maintain the relative proportions of the two assets the AMM already holds. |
+| `Asset1In` and `Asset2In` | Deposit both of this AMM's assets, up to the specified amounts. The actual amounts deposited must maintain the same balance of assets as the AMM already holds, so the amount of either one deposited MAY be less than specified. The amount of LP Tokens you get in return is based on the total value deposited. |
```
Flag `tfTwoAsset`:0x04

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,109 @@
+
+| Fields Specified             | Meaning |
+|------------------------------|---------|
+| `Asset1In` only              | Deposit exactly the specified amount of one asset, and receive an amount of LP Tokens based on the resulting share of the pool (minus fees). |
```
Flag `tfSingleAssset`: 0x02

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,109 @@
+| Fields Specified             | Meaning |
+|------------------------------|---------|
+| `Asset1In` only              | Deposit exactly the specified amount of one asset, and receive an amount of LP Tokens based on the resulting share of the pool (minus fees). |
+| `Asset1In` and `LPTokensOut` | Deposit up to the specified amount of one asset, so that you receive exactly the specified amount of LP Tokens in return (after fees). |
```
Flag `tfOneAssetLPToken`: 0x08

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,109 @@
+|------------------------------|---------|
+| `Asset1In` only              | Deposit exactly the specified amount of one asset, and receive an amount of LP Tokens based on the resulting share of the pool (minus fees). |
+| `Asset1In` and `LPTokensOut` | Deposit up to the specified amount of one asset, so that you receive exactly the specified amount of LP Tokens in return (after fees). |
+| `Asset1In` and `EPrice`      | Deposit up to the specified amount of one asset, but pay no more than the specified effective price per LP Token (after fees). |
```
Flag `tfLimitLPToken`:0x10

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammvote.md:
````diff
@@ -0,0 +1,54 @@
+```json
+{
+  "TransactionType": "AMMVote",
+  "AMMID": "4B82E4CB90FBE26EA172A498F7A7B03A2A7C285149A9B18731370F2644B96F7A",
````
Replaced with `Asset/Asset2` as in other tx.

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,99 @@
+  "TransactionType": "AMMWithdraw",
+  "AMMID": "4B82E4CB90FBE26EA172A498F7A7B03A2A7C285149A9B18731370F2644B96F7A",
+  "Account": "rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm",
+  "Asset1Out": {
```
Replaced with `Amount/Amount2`

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,99 @@
+| Fields Specified                           | Meaning |
+|--------------------------------------------|---------|
+| No fields; `tfAMMWithdrawAll` flag enabled | Return _all_ of your LP Tokens and receive as much as you can of both assets in the AMM's pool. |
+| `LPTokensIn` only                          | Return the specified amount of LP Tokens and receive both assets from the AMM's pool in amounts based on the returned LP Tokens' share of the total LP Tokens issued. |
```
Flag: `tfLPToken`:0x01

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,99 @@
+|--------------------------------------------|---------|
+| No fields; `tfAMMWithdrawAll` flag enabled | Return _all_ of your LP Tokens and receive as much as you can of both assets in the AMM's pool. |
+| `LPTokensIn` only                          | Return the specified amount of LP Tokens and receive both assets from the AMM's pool in amounts based on the returned LP Tokens' share of the total LP Tokens issued. |
+| `Asset1Out` and `Asset2Out`                | Withdraw both of this AMM's assets, in up to the specified amounts. The actual amounts received maintains the balance of assets in the AMM's pool. |
```
Flag `tfTwoAsset`:0x04

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,99 @@
+
+| Fields Specified                             | Meaning |
+|----------------------------------------------|---------|
+| `Asset1Out` only                             | Withdraw exactly the specified amount of one asset, by returning as many LP Tokens as necessary. |
```
Flag `tfSingleAsset`:0x02.

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,99 @@
+|----------------------------------------------|---------|
+| `Asset1Out` only                             | Withdraw exactly the specified amount of one asset, by returning as many LP Tokens as necessary. |
+| `Asset1Out`; `tfAMMWithdrawAll` flag enabled | Withdraw at least the specified amount of one asset, by returning _all_ of your LP Tokens. (Fails if you can't receive at least the specified amount.) |
+| `Asset1Out` and `LPTokensIn`                 | Withdraw up to the specified amount of one asset, by returning up to the specified amount of LP Tokens. |
```
Flag `tfOneAssetLPToken`: 0x08

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,99 @@
+| `Asset1Out` only                             | Withdraw exactly the specified amount of one asset, by returning as many LP Tokens as necessary. |
+| `Asset1Out`; `tfAMMWithdrawAll` flag enabled | Withdraw at least the specified amount of one asset, by returning _all_ of your LP Tokens. (Fails if you can't receive at least the specified amount.) |
+| `Asset1Out` and `LPTokensIn`                 | Withdraw up to the specified amount of one asset, by returning up to the specified amount of LP Tokens. |
+| `Asset1Out` and `EPrice`                     | Withdraw up to the specified amount of one asset, but pay no more than the specified effective price in LP Tokens per unit of the asset received. |
```
Flag `tfLimitLPToken`: 0x10

gregtatcam commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,99 @@
+
+| Flag Name             | Hex Value    | Decimal Value | Description           |
+|:----------------------|:-------------|:--------------|:----------------------|
+| `tfAMMWithdrawAll`    | `0x00010000` | 65536         | Withdraw as much as possible by returning all your LP Tokens. |
```
Changed to `tfWithdrawAll`: 0x20

gregtatcam reviewed (changes requested):
Thanks, looks good! I have only paid attention mostly to the naming convention so far. I'll take another pass later. I will be updating the error codes in the next commit (there is a ticket for the error code consolidation) and I want to review the errors code in general. I'll provide the update before I push the next commit.

jazzi-cooper commented on content/references/protocol-reference/ledger-data/ledger-object-types/amm.md:
```diff
@@ -0,0 +1,138 @@
+| `AMMToken`       | Object              | STObject          | Yes       | Defines the two assets this AMM trades, as an [AMMToken Object](#ammtoken-object). |
+| `AccountRootID`  | String              | AccountID         | Yes       | The address of the special account that holds this AMM's assets. |
+| `AuctionSlot`    | Object              | STObject          | No        | Details of the current owner of the auction slot, as an [Auction Slot object](#auction-slot-object). |
+| `LPTokenBalance` | [Currency Amount][] | Amount            | Yes       | The total outstanding balance of liquidity provider tokens from this AMM instance. The holders of these tokens can vote on the AMM's trading fee in proportion to their holdings, or redeem the tokens for a share of the AMM's assets. |
```
This phrasing excludes the fact that LP token holders are earning trading fees while they remain in the pool (which is the primary purpose of participating), in addition to their voting rights. 

When they redeem the LP tokens, they are redeeming their original contribution plus trading fees earned.

mDuo13 replied:
Yeah, I mean, the part where you can redeem your tokens for a share of the AMM's assets was meant to encompass that implicitly, but I rewrote to make it explicit.

jazzi-cooper commented on content/references/protocol-reference/ledger-data/ledger-object-types/amm.md:
```diff
@@ -0,0 +1,138 @@
+| `AccountRootID`  | String              | AccountID         | Yes       | The address of the special account that holds this AMM's assets. |
+| `AuctionSlot`    | Object              | STObject          | No        | Details of the current owner of the auction slot, as an [Auction Slot object](#auction-slot-object). |
+| `LPTokenBalance` | [Currency Amount][] | Amount            | Yes       | The total outstanding balance of liquidity provider tokens from this AMM instance. The holders of these tokens can vote on the AMM's trading fee in proportion to their holdings, or redeem the tokens for a share of the AMM's assets. |
+| `TradingFee`     | Number              | UInt16            | Yes       | The percentage fee to be charged for trades against this AMM instance, in units of 1/10,000. The maximum value is 65000, for a 65% fee. |
```
The fee max will drop from 65% to 1%, this should be reflected in latest spec

mDuo13 committed:
[JA] add untranslated section on AMMs

mDuo13 commented on content/references/protocol-reference/ledger-data/ledger-object-types/accountroot.md:
```diff
@@ -57,8 +57,9 @@ The `AccountRoot` object has the following fields:
 | `TransferRate`                | Number    | UInt32            | No        | A [transfer fee](transfer-fees.html) to charge other users for sending currency issued by this account to each other. |
-| `WalletLocator`               | String    | Hash256           | No        | **DEPRECATED**. Do not use. |
-| `WalletSize`                  | Number    | UInt32            | No        | **DEPRECATED**. Do not use. |
+| `WalletLocator`               | String    | Hash256           | No        | An arbitrary 256-bit value that users can set. |
+| `WalletSize`                  | Number    | UInt32            | No        | Unused. (The code supports this field but there is no way to set it.) |
```
These changes are unrelated to AMM; they match some work from #1517 which updated the transaction but not the object definition. I did them "while I was here" since I was already modifying this page.

mDuo13 committed:
AMM doc cleanup

- fix typo in serialization
- add STIssue type to type list in serialization
- more clearly mark things that need to be translated to Japanese
- Put AMMBid & AMMCreate transactions in alphabetical order

mDuo13 committed:
[JA] serialization, fix link

oeggert commented on content/_snippets/amm-disclaimer.md:
```diff
@@ -0,0 +1 @@
+_Automated Market Maker (AMM) functionality is part of the proposed [XLS-30d](https://github.com/XRPLF/XRPL-Standards/discussions/78) extension :not_enabled: to the XRP Ledger protocol. You can use these functions on AMM test networks, but there is not yet an official amendment and they are not yet available on the production Mainnet. Until there is an amendment, the details documented on these p ...
```
```suggestion
_Automated Market Maker (AMM) functionality is part of the proposed [XLS-30d](https://github.com/XRPLF/XRPL-Standards/discussions/78) extension :not_enabled: to the XRP Ledger protocol. You can use these functions on AMM test networks, but there isn't an official amendment and they aren't available on the production Mainnet. Until there is an amendment, the details documented on these pages are subject to change._
```
Minor change to match style guide on contractions.

oeggert commented on content/references/protocol-reference/ledger-data/ledger-object-types/accountroot.ja.md:
```diff
@@ -56,6 +56,22 @@ labels:
+
+{% include '_snippets/amm-disclaimer.md' %}
+
+Automated Market Makers (AMMs) use an AccountRoot object to issue their LP Tokens and hold the assets in the AMM pool, in addition to the [AMM object][] for tracking some of the details of the AMM. The address of an AMM's associated AccountRoot is randomized so that users cannot identify and fund the address in advance of the AMM being created. Unlike normal accounts, AMM AccountRoots are created ...
```
Do we need to make mention of the AMM object here? It has its own page going over that info.

mDuo13 replied:
I think it's useful to mention it here in case people read this one first—just so people know that an AMM is always two ledger entries, one of each type.

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,136 @@
+
+{% include '_snippets/amm-disclaimer.md' %}
+
+Bid on an Automated Market Maker's (AMM's) auction slot. If this transaction succeeds, you win the bid and can trade against the AMM at a discounted fee until you are outbid or 24 hours have passed. If you are outbid before 24 hours have passed, you are refunded part of the cost of your bid based on how much time remains.
```
```suggestion
Bid on an Automated Market Maker's (AMM's) auction slot. If you win, you can trade against the AMM at a discounted fee until you are outbid or 24 hours have passed. If you are outbid before 24 hours have passed, you are refunded part of your bid based on how much time remains.
```
Can we remove explicitly mentioning this is a transaction since we're already under transaction types in page hierarchy.

mDuo13 replied:
Hm, I'm not sure how I feel about this suggestion. I like that it's more succinct. My thought behind the original phrasing is that I want to answer the questions, "how do you know if you won?" or "what does it mean for this transaction to be successful?" since there are cases where "success" at the level of executing the transaction doesn't necessarily mean what you might think. (For example, when placing an Offer in the DEX, "success" might mean, "your Offer has been recorded and is waiting for a match", not "you successfully traded currency". Or when canceling a prior Offer, "success" means "there is no longer any such Offer" which can mean "it has been canceled" or "it was already fully consumed" as well as "it never existed because you made a typo".) So I do think it's useful to convey, "if the transaction has a success result code, that _actually_ means you won the bid" and not just, "your bid was received, but it may or may not have been too low to win.")

oeggert replied:
Ah okay. The distinction on what a successful transaction means makes sense to include.

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,136 @@
+
+**Note:** To make sure all servers in the network reach the same results when building a ledger, time measurements are based on the [official close time](ledgers.html#ledger-close-times) of the previous ledger, which is approximate.
+
+## Refunds
```
Maybe rename  this header to specify to Bid Refunds or Auction Slot Refunds?

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,136 @@
+
+## Refunds
+
+In the case where the transaction outbids the previous slot holder, the AMM refunds the previous holder part of the price of the auction slot. The amount refunded is calculated using the following formula:
```
```suggestion
When a transaction outbids an auction slot holder, the AMM refunds the previous holder using this formula:
```
Just a suggestion for brevity.

legleux commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,116 @@
+| `Amount`     | [Currency Amount][] | Amount            | No        | The amount of one asset to withdraw from the AMM. This must match the type of one of the assets (tokens or XRP) in the AMM's pool. |
+| `Amount2`    | [Currency Amount][] | Amount            | No        | The amount of another asset to withdraw from the AMM. If present, this must match the type of the other asset in the AMM's pool and cannot be the same type as `Amount`. |
+| `EPrice`     | [Currency Amount][] | Amount            | No        | The minimum effective price, in LP Token returned, to pay per unit of the asset to withdraw. |
+| `LPTokensIn` | [Currency Amount][] | Amount            | No        | How many of the AMM's LP Tokens to redeem. |
```
LPTokensIn --> LPTokenIn
idk, maybe LPTokenOut should be LPTokensOut.

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,136 @@
+- `B` is the price of the previous bid to be refunded, in LP Tokens.
+- `t` is the fraction of time elapsed in the current 24-hour slot, rounded down to a multiple of 0.05.
+
+As a special case, during the final (20th) interval of the auction slot, the refunded amount is zero.
```
```suggestion
**Note:** No refund is issued during the last 72 minutes of an auction slot.
```

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,136 @@
+
+Like other "inner objects" that can appear in arrays, the JSON representation of each of these objects is wrapped in an object whose only key is the object type, `AuthAccount`.
+
+## Slot Price Calculations
```
Same suggestion as the one on refunds. Maybe rename this to Auction Slot Prices?

legleux commented:
Hmm, `a {{currentpage.name}}` renders to `a AMM<txn>`
I guess that's everywhere. Probably can't do anything about it easily?

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,136 @@
+
+## Slot Price Calculations
+
+If successful, the transaction automatically outbids the previous slot owner and debits the bid price from the sender's LP Tokens. The price to win the auction decreases over time, divided into 20 intervals of 72 minutes each. If the sender does not have enough LP Tokens to win the bid, or the price of the bid is higher than the transaction's `BidMax` value, the transaction fails with a `tecAMM_F ...
```
I might be missing something, but it doesn't seem like the 20 intervals detail comes into play except in the first and last interval. As written I assumed the current price decrease happens at each 72 minute interval, but the formula used just takes a fraction of 24 hour periods.

I've made several suggestions to maybe simplify this and just say the first/last 72 minutes. Feel free to disregard if the interval thing is important.

mDuo13 replied:
My understanding is that the formula uses a _rounded fraction_ of 24 hour periods. The rounding means that the price increase _does_ happen at each interval.

Also, the `amm_info` method reports what interval the current winning bid is in, so the intervals concept is useful to understand for that purpose as well, I think.

I'm going to think more on this to decide which way to go on the intervals thing.

oeggert replied:
I wasn't aware of the amm_info method reporting in intervals, in which case I would leave as is.

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,136 @@
+
+- Otherwise, the price to outbid the current holder is calculated using the following formula:
+
+        P = B × 1.05 × (1 - t⁶⁰) + M
```
Am I understanding this right? t⁶⁰ is elapsed time to the 60th power?

mDuo13 replied:
line 91: elapsed fraction of time, rounded to a multiple of 0.05, and then raised to the 60th power. So for example in the third interval, t=0.15 therefore t⁶⁰ = `3.67e-50` (where `e-50` means "times 10 raised to the negative-50th power") Something like that.

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,136 @@
+
+If successful, the transaction automatically outbids the previous slot owner and debits the bid price from the sender's LP Tokens. The price to win the auction decreases over time, divided into 20 intervals of 72 minutes each. If the sender does not have enough LP Tokens to win the bid, or the price of the bid is higher than the transaction's `BidMax` value, the transaction fails with a `tecAMM_F ...
+
+- If the auction slot is currently empty, expired, or in its last interval, the **minimum bid** is **0.001% of the AMM's total LP Tokens balance**.
```
```suggestion
- If the auction slot is empty, expired, or in its last 72 minutes, the **minimum bid** is **0.001% of the AMM's total LP Tokens balance**.
```

mDuo13 replied:
Decided after thinking about it more not to change the intervals thing.

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,136 @@
+    - `t` is the fraction of time elapsed in the current 24-hour slot, rounded down to a multiple of 0.05.
+    - `M` is the **minimum bid** as defined above.
+
+    There are two special cases for the cost to outbid someone. In the **first interval** after someone wins the bid, the price to outbid them is the minimum bid plus 5% more than the existing bid:
```
```suggestion
There are two special cases for the cost to outbid someone. In the first 72 minutes after someone wins the bid, the price to outbid them is the minimum bid plus 5% more than the existing bid:
```

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammbid.md:
```diff
@@ -0,0 +1,136 @@
+
+        P = B × 1.05 + M
+
+    In the **last interval** of someone's slot, the cost to outbid someone is only the minimum bid:
```
```suggestion
In the last 72 minutes of someone's slot, the cost to outbid someone is only the minimum bid:
```

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammcreate.md:
```diff
@@ -0,0 +1,71 @@
+
+{% include '_snippets/amm-disclaimer.md' %}
+
+Create a new Automated Market-Maker (AMM) instance for trading a given pair of assets ([fungible tokens](tokens.html) or [XRP](xrp.html)).
```
```suggestion
Create a new Automated Market-Maker (AMM) instance for trading a pair of assets ([fungible tokens](tokens.html) or [XRP](xrp.html)).
```

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammcreate.md:
```diff
@@ -0,0 +1,71 @@
+
+Create a new Automated Market-Maker (AMM) instance for trading a given pair of assets ([fungible tokens](tokens.html) or [XRP](xrp.html)).
+
+If successful, this transaction creates both an [AMM object][] and a [special AccountRoot object](accountroot.html#special-amm-accountroot-objects) which, together, represent the AMM; it also transfers ownership of the starting balance of both assets from the sender to the created `AccountRoot`, and issues an initial balance of liquidity provider tokens (LP Tokens) from the AMM account to this tr ...
```
```suggestion
If successful, this transaction creates an [AMM object][] and a [special AccountRoot object](accountroot.html#special-amm-accountroot-objects) to represent the AMM. It also transfers ownership of the starting balance of both assets from the sender to the created `AccountRoot` and issues an initial balance of liquidity provider tokens (LP Tokens) from the AMM account to the sender.
```

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,131 @@
+
+{% include '_snippets/amm-disclaimer.md' %}
+
+Deposit funds into an Automated Market-Maker (AMM) instance and receive the AMM's liquidity provider tokens (_LP Tokens_) in exchange. You can deposit one or both of the assets in the AMM's pool.
```
```suggestion
Deposit funds into an Automated Market-Maker (AMM) instance and receive the AMM's liquidity provider tokens (_LP Tokens_) in exchange. You can deposit either asset in the AMM's pool.
```

mDuo13 replied:
The intended meaning is you can deposit A, B, or A+B (and the semantics of A+B are different than depositing just one). Suggested wording sounds like the options are A or B (single-asset deposit) only.

oeggert replied:
Misread the original. Keep as is.

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,131 @@
+
+### AMMDeposit Modes
+
+This transaction has several modes, depending on which flags you specify. Each mode expects a specific combination of fields. The modes fall into two categories: 
```
```suggestion
This transaction has two deposit modes:
```

mDuo13 replied:
The transaction has 5 modes, two of which are double-asset deposits and three of which are single-asset deposits. Each mode has a specific subset of allowed fields _and_ works differently in terms of how it decides how much money to deposit (where it places the limits). As much as I would like to say it has 2 modes, I think that ends up oversimplifying.

oeggert replied:
I misinterpreted the original meaning, then. Disregard.

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,131 @@
+
+This transaction has several modes, depending on which flags you specify. Each mode expects a specific combination of fields. The modes fall into two categories: 
+
+- **Double-asset deposits**, in which you provide both assets in the AMM's pool in proportions that match the balance of those assets already there. These deposits are not subject to a fee.
```
- **Double-asset deposits**, in which you provide both assets in the AMM's pool, proportional to the balance of the assets already there. These deposits are not subject to a fee.

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,131 @@
+This transaction has several modes, depending on which flags you specify. Each mode expects a specific combination of fields. The modes fall into two categories: 
+
+- **Double-asset deposits**, in which you provide both assets in the AMM's pool in proportions that match the balance of those assets already there. These deposits are not subject to a fee.
+- **Single-asset deposits**, in which you provide only one of the AMM's two assets. You can envision a single-asset deposit as trading _some_ of the deposited asset for the other asset, then performing a double-asset deposit. The trading fee applies to the amount you would need to trade for, but not to the rest of the deposit; the amount of the fee is debited from the LP Tokens paid out. _For exa ...
```
The way this is worded is confusing. I don't think we should use the example of trading one asset for another, since it implies you're getting a withdrawal of the other asset. Maybe we elaborate on how the AMM will try to balance the difference, and in doing so subract trading fees from LP token payouts as it tries to balance the values.

mDuo13 replied:
The way this _works_ is confusing; this wording is at least less confusing to me than anything else I've come up with. 

I'm not sure what you mean by elaborating on "how the AMM will try to balance the difference" instead. The AMM doesn't move anything around to try and make things balance, it just charges the trading fee on a percentage of your deposit. And it was helpful to me at least to know that it's calculated so that the result is identical to if you did a trade against the AMM first and then did a double-asset deposit.

mDuo13 replied:
Changed to this:

> - **Single-asset deposits**, in which you provide only one of the AMM's two assets. The AMM charges a fee, debited from the LP Tokens paid out, based on how much your deposit shifts the balance of assets in the pool.

Then I moved some details to the Single Asset Deposit Fee Calculations section below:

> ### Single Asset Deposit Fee Calculations
>
> The fee for a single asset deposit is calculated to be the same as if you had used the AMM to trade part of the deposit amount for the other asset, then performed a double-asset deposit. The AMM's trading fee applies to the amount you would need to trade for, but not to the rest of the deposit. _For example, if the AMM's asset pool is split perfectly evenly between USD and EUR, and you try to deposit 100 USD, the amount of LP Tokens you receive is slightly less than if you had deposited 50 EUR + 50 USD, because you pay the trading fee to convert some of your USD to an equal amount of EUR._

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:
```diff
@@ -0,0 +1,131 @@
+
+### Single Asset Deposit Fee Calculations
+
+If you deposit one asset, the AMM charges a fee by reducing the total amount of LP Tokens it pays out, weighted by how much your deposit shifts the balance of assets in the pool. This is equivalent to if you had traded part of the deposit amount for the other asset, then performed a double-asset deposit. The formula for how many LP Tokens you receive for a double-asset deposit is as follows:
```
If we make the explanation for single asset deposits more concise in the previous section, we won't have to reiterate what it is here and can instead go straight into the fee calculation.

mDuo13 replied:
Tweaked this section based on your feedback (see comment above for what I went with)

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,116 @@
+
+{% include '_snippets/amm-disclaimer.md' %}
+
+Withdraw funds an Automated Market-Maker (AMM) instance by returning the AMM's liquidity provider tokens (LP Tokens).
```
```suggestion
Withdraw assets from an Automated Market-Maker (AMM) instance by returning the AMM's liquidity provider tokens (LP Tokens).
```
I think it's better to be precise here with asset withdrawal.

oeggert commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,116 @@
+
+### AMMWithdraw Modes
+
+This transaction has several modes, depending on which flags you specify. Each mode expects a specific combination of fields. The modes fall into two categories: 
```
```suggestion
This transaction has two withdrawal modes:
```

oeggert reviewed (commented):
Most of these are suggestions for brevity and clarity, but I do have a couple technical questions to verify.

mDuo13 committed:
AMM: apply suggestions from review

Co-authored-by: oeggert <117319296+oeggert@users.noreply.github.com>

mDuo13 committed:
AMM refs: more edits per reviews

mDuo13 commented:
Rebased on master with all the latest changes so there shouldn't be a merge conflict anymore.

jazzi-cooper commented on content/references/protocol-reference/transactions/transaction-types/ammwithdraw.md:
```diff
@@ -0,0 +1,116 @@
+| `Asset2`     | Object              | STIssue           | Yes       | The definition for the other asset in the AMM's pool. In JSON, this is an object with `currency` and `issuer` fields (omit `issuer` for XRP). |
+| `Amount`     | [Currency Amount][] | Amount            | No        | The amount of one asset to withdraw from the AMM. This must match the type of one of the assets (tokens or XRP) in the AMM's pool. |
+| `Amount2`    | [Currency Amount][] | Amount            | No        | The amount of another asset to withdraw from the AMM. If present, this must match the type of the other asset in the AMM's pool and cannot be the same type as `Amount`. |
+| `EPrice`     | [Currency Amount][] | Amount            | No        | The minimum effective price, in LP Token returned, to pay per unit of the asset to withdraw. |
```
Is this definition of effective price correct? My understanding is that it is calculated as the ratio of the tokens swapped in vs out, giving the "effective price" of the trade. I don't think LP token quantity is a factor in the equation, but maybe i'm wrong and that works too

jazzi-cooper replied:
Sorry, I'm wrong because of the txn type (withdraw, not swap). Ignore the above

mDuo13 replied:
In the case of a withdraw with `EPrice`, what's swapped in _is_ LP Tokens

mDuo13 committed:
Fix AMMDeposit formula
